strawberryfields/engine.py

```diff
@@ -223,11 +225,16 @@ def _run(self, program, *, compile_options={}, **kwargs):
 
                 # if the program hasn't been compiled for this backend, do it now
                 if p.backend != self.backend_name:
-                    p = p.compile(self.backend_name, **compile_options)
+                    p = p.compile(self.backend_name, **compile_options) # TODO: shots might be relevant for compilation?
                 p.lock()
 
+                kwargs["shots"] = shots
+                # Note: by putting ``shots`` into keyword arguments, it allows for the
+                # signatures of methods in Operations to remain cleaner, since only
+                # Measurements need to know about shots
+
                 self._run_program(p, **kwargs)
-                self.run_progs.append(p)
+                self.run_progs.append(p) #TODO: is the number of shots something that should be stored in ``p``?
```

Comment:
@smite @josh146 for discussion

Reply:
IMO shots should have nothing to do with the Program class. `Program` describes a quantum circuit or a fragment thereof, it should not concern itself on how it is used.

Reply:
seems like the smartest approach, just want to make sure there are no edge-cases